test/integration/thruster.cc

```diff
@@ -0,0 +1,216 @@
+
+      auto propellerEntity = model.LinkByName(_ecm, "propeller");
+      EXPECT_NE(propellerEntity, kNullEntity);
+
+      propeller = Link(propellerEntity);
+      propeller.EnableVelocityChecks(_ecm);
+    }).
+  OnPostUpdate([&](const gazebo::UpdateInfo &_info,
+                            const gazebo::EntityComponentManager &_ecm)
+    {
+      dt = std::chrono::duration_cast<std::chrono::nanoseconds>(
+          _info.dt).count() * 1e-9;
+
+      auto modelPose = worldPose(model.Entity(), _ecm);
+      modelPoses.push_back(modelPose);
+
+      auto propellerAngVel = propeller.WorldAngularVelocity(_ecm);
+      ASSERT_TRUE(propellerAngVel);
+      propellerAngVels.push_back(propellerAngVel.value());
+    }).
+  Finalize();
+
+  // Check initial position
+  fixture.Server()->Run(true, 100, false);
+  EXPECT_EQ(100u, modelPoses.size());
+  EXPECT_EQ(100u, propellerAngVels.size());
+
+  EXPECT_NE(model.Entity(), kNullEntity);
+  EXPECT_NE(propeller.Entity(), kNullEntity);
+
+  for (const auto &pose : modelPoses)
+  {
+    EXPECT_EQ(math::Pose3d(), pose);
+  }
+  modelPoses.clear();
+  for (const auto &vel : propellerAngVels)
+  {
+    EXPECT_EQ(math::Vector3d::Zero, vel);
+  }
+  propellerAngVels.clear();
+
+  // Publish command and check that vehicle moved
+  transport::Node node;
+  auto pub = node.Advertise<msgs::Double>(
+      "/model/" + _namespace + "/joint/propeller_joint/cmd_thrust");
+
+  int sleep{0};
+  int maxSleep{30};
+  for (; !pub.HasConnections() && sleep < maxSleep; ++sleep)
+  {
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
+  }
+  EXPECT_LT(sleep, maxSleep);
+  EXPECT_TRUE(pub.HasConnections());
+
+  double force{300.0};
+  msgs::Double msg;
+  msg.set_data(force);
+  pub.Publish(msg);
+
+  // Check movement
+  for (sleep = 0; modelPoses.back().Pos().X() < 5.0 && sleep < maxSleep;
+      ++sleep)
+  {
+    std::this_thread::sleep_for(std::chrono::milliseconds(100));
+    fixture.Server()->Run(true, 100, false);
+  }
+  EXPECT_LT(sleep, maxSleep);
+  EXPECT_LT(5.0, modelPoses.back().Pos().X());
+
+  EXPECT_EQ(100u * sleep, modelPoses.size());
+  EXPECT_EQ(100u * sleep, propellerAngVels.size());
+
+  // F = m * a
+  // s = a * t^2 / 2
+  // F = m * 2 * s / t^2
+  // s = F * t^2 / 2m
+  double mass{100.1};
+  double xTol{1e-2};
+  for (unsigned int i = 0; i < modelPoses.size(); ++i)
+  {
+    auto pose = modelPoses[i];
+    auto time = dt * i;
+    EXPECT_NEAR(force * time * time / (2 * mass), pose.Pos().X(), xTol);
+    EXPECT_NEAR(0.0, pose.Pos().Y(), _baseTol);
+    EXPECT_NEAR(0.0, pose.Pos().Z(), _baseTol);
+    EXPECT_NEAR(0.0, pose.Rot().Pitch(), _baseTol);
+    EXPECT_NEAR(0.0, pose.Rot().Yaw(), _baseTol);
+
+    // The joint velocity command adds some roll to the body which the PID
+    // wrench doesn't
+    if (_namespace == "custom")
+      EXPECT_NEAR(0.0, pose.Rot().Roll(), 0.1);
+    else
+      EXPECT_NEAR(0.0, pose.Rot().Roll(), _baseTol);
+  }
+
+  // omega = sqrt(thrust /
```

Comment:
Should we add the book reference in the test for clarity?

Reply:
9a0642b